brainzutils/ratelimit.py

```diff
@@ -0,0 +1,261 @@
+
+          def validate_user(user):
+              if user == valid_user:
+                  return True
+              return False
+
+         set_user_validation_function(validate_user)
+
+'''
+
+
+class RateLimit(object):
+    '''
+        This Ratelimit object is created when a request is started (via the ratelimit decorator)
+        and is stored in the flask's request context so that the results can be injected into
+        the response headers before the request is over.
+    '''
+
+    # From the docs:
+    # We also give the key extra expiration_window seconds time to expire in cache so that badly
+    # synchronized clocks between the workers and the cache server do not cause problems.
+    expiration_window = 10
+
+    def __init__(self, key_prefix, limit, per):
+        current_time = int(time.time())
+        self.reset = (current_time // per) * per + per
+        self.seconds_before_reset = self.reset - current_time
+        self.key = key_prefix + str(self.reset)
+        self.limit = limit
+        self.per = per
+        self.current = cache.increment(self.key)
+        cache.expireat(self.key, self.reset + self.expiration_window)
+
+    remaining = property(lambda x: max(x.limit - x.current, 0))
+    over_limit = property(lambda x: x.current > x.limit)
+
+
+def set_user_validation_function(func):
+    '''
+        The function passed to this method should accept on argument, the Authorization header contents
+        and return a True/False value if this user is a valid user.
+    '''
+
+    global ratelimit_user_validation
+    ratelimit_user_validation = func
+
+
+def set_rate_limits(per_token, per_ip, window):
+    '''
+        Update the current rate limits. This will affect all new rate limiting windows and existing windows will not be changed.
+    '''
+    cache.set(ratelimit_per_token_key, per_token)
+    cache.set(ratelimit_per_ip_key, per_ip)
+    cache.set(ratelimit_window_key, window)
+
+
+def inject_x_rate_headers(response):
+    '''
+        Add rate limit headers to responses
+    '''
+    limit = get_view_rate_limit()
+    if limit:
+        h = response.headers
+        h.add('Access-Control-Expose-Headers', 'X-RateLimit-Remaining,X-RateLimit-Limit,X-RateLimit-Reset,X-RateLimit-Reset-In')
+        h.add('X-RateLimit-Remaining', str(limit.remaining))
+        h.add('X-RateLimit-Limit', str(limit.limit))
+        h.add('X-RateLimit-Reset', str(limit.reset))
+        h.add('X-RateLimit-Reset-In', str(limit.seconds_before_reset))
+    return response
+
+
+def get_view_rate_limit():
+    '''
+        Helper function to fetch the ratelimit limits from the flask context
+    '''
+    return getattr(g, '_view_rate_limit', None)
+
+
+def on_over_limit(limit):
+    ''' 
+        Set a nice and readable error message for over the limit requests.
+    '''
+    return 'You have exceeded your rate limit. See the X-RateLimit-* response headers for more ' \
```

Comment:
While clients should check status codes, it's possible that they just directly try and interpret all requests as a specific format (json?). What about letting an app specify the format that this error message is returned in?